Source/Core/Core/IOS/Network/KD/VFF/VFFUtil.cpp

```diff
@@ -0,0 +1,297 @@
+  }
+  else if (cluster_count < 65525)
+  {
+    fs->fs_type = FS_FAT16;
+
+    u32 table_size = cluster_count * 2;
+    table_size = (table_size + cluster_size - 1) & ~(cluster_size - 1);
+    fs->fsize = table_size / 512;
+  }
+  else
+  {
+    ERROR_LOG_FMT(IOS_WC24, "VFF not FAT12 or 16! Cluster size: {}", cluster_size);
+    return RES_ERROR;
+  }
+
+  fs->n_fats = 2;
+  fs->csize = 1;
+
+  // Root directory entry is 4096 bytes long, with each entry being 32 bytes. 4096 / 32 = 128
+  fs->n_rootdir = 128;
+
+  u32 sysect = 1 + (fs->fsize * 2) + fs->n_rootdir / (512 / 32);
+
+  // cluster_count is the total count whereas this is the actual amount of clusters we can use
+  u32 actual_cluster_count = cluster_count - sysect;
+
+  fs->n_fatent = actual_cluster_count + 2;
+  fs->volbase = 0;
+  fs->fatbase = 1;
+  fs->database = sysect;
+  // Root directory entry
+  fs->dirbase = fs->fatbase + fs->fsize * 2;
+
+  // Initialize cluster allocation information
+  fs->last_clst = fs->free_clst = 0xFFFFFFFF;
+  fs->fsi_flag = 0x80;
+
+  fs->id = 0;
+  fs->cdir = 0;
+
+  return RES_OK;
+}
+
+DRESULT vff_read(BYTE pdrv, BYTE* buff, LBA_t sector, UINT count)
+{
+  const u64 offset = static_cast<u64>(sector) * SECTOR_SIZE - 480;
+  if (!s_vff->Seek(offset, IOS::HLE::FS::SeekMode::Set))
+  {
+    ERROR_LOG_FMT(IOS_WC24, "VFF seek failed (offset={})", offset);
+    return RES_ERROR;
+  }
+
+  const size_t size = static_cast<size_t>(count) * SECTOR_SIZE;
+  if (!s_vff->Read(buff, size))
+  {
+    ERROR_LOG_FMT(IOS_WC24, "VFF read failed (offset={}, size={})", offset, size);
+    return RES_ERROR;
+  }
+
+  return RES_OK;
+}
+
+DRESULT vff_write(BYTE pdrv, const BYTE* buff, LBA_t sector, UINT count)
+{
+  const u64 offset = static_cast<u64>(sector) * SECTOR_SIZE - 480;
+  if (!s_vff->Seek(offset, IOS::HLE::FS::SeekMode::Set))
+  {
+    ERROR_LOG_FMT(IOS_WC24, "VFF seek failed (offset={})", offset);
+    return RES_ERROR;
+  }
+
+  const size_t size = static_cast<size_t>(count) * SECTOR_SIZE;
+  const auto res = s_vff->Write(buff, size);
+  if (!res)
+  {
+    ERROR_LOG_FMT(IOS_WC24, "VFF write failed (offset={}, size={})", offset, size);
+    return RES_ERROR;
+  }
+
+  return RES_OK;
+}
+
+DRESULT vff_ioctl(BYTE pdrv, BYTE cmd, void* buff)
+{
+  switch (cmd)
+  {
+  case CTRL_SYNC:
+    return RES_OK;
+  case GET_SECTOR_COUNT:
+    *reinterpret_cast<LBA_t*>(buff) = s_vff->GetStatus()->size / SECTOR_SIZE;
+    return RES_OK;
+  default:
+    WARN_LOG_FMT(IOS_WC24, "Unexpected FAT ioctl {}", cmd);
+    return RES_OK;
+  }
+}
+
+namespace IOS::HLE::NWC24
+{
+static ErrorCode WriteFile(const std::string& filename, const std::vector<u8>& tmp_buffer)
+{
+  FIL dst;
+  const auto open_error_code = f_open(&dst, filename.c_str(), FA_CREATE_ALWAYS | FA_WRITE);
+  if (open_error_code != FR_OK)
+  {
+    ERROR_LOG_FMT(IOS_WC24, "Failed to open file {} in VFF", filename);
+    return WC24_ERR_FILE_OPEN;
+  }
+
+  // FIXME: Handle written_size < tmp_buffer.size()
```

Comment:
FYI, I added this FIXME. I'm not sure if this ever happens in practice, but it's not too hard to add (see `Pack()` in `FatFsUtil.cpp` for a reference), so this probably should.